conjure-runtime/src/service/timeout.rs

```diff
@@ -74,154 +86,177 @@ where
 pub struct TimeoutFuture<F> {
     #[pin]
     future: F,
-    #[pin]
-    delay: Delay,
+    read_timeout: Duration,
+    write_timeout: Duration,
 }
 
-impl<F, B> Future for TimeoutFuture<F>
+impl<F, S, E> Future for TimeoutFuture<F>
 where
-    F: Future<Output = Result<Response<B>, Error>>,
+    F: Future<Output = Result<S, E>>,
+    S: AsyncRead + AsyncWrite + Unpin,
 {
-    type Output = Result<Response<TimeoutBody<B>>, Error>;
+    type Output = Result<TimeoutStream<S>, E>;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let mut this = self.project();
-
-        if this.delay.as_mut().poll(cx).is_ready() {
-            return Poll::Ready(Err(Error::internal_safe(TimeoutError(()))));
-        }
+        let this = self.project();
 
-        let response = ready!(this.future.poll(cx))?;
+        let stream = ready!(this.future.poll(cx))?;
+        let mut stream = tokio_io_timeout::TimeoutStream::new(stream);
+        stream.set_read_timeout(Some(*this.read_timeout));
+        stream.set_write_timeout(Some(*this.write_timeout));
 
-        let deadline = this.delay.deadline();
-        Poll::Ready(Ok(response.map(|body| TimeoutBody {
-            body,
-            delay: time::delay_until(deadline),
-        })))
+        Poll::Ready(Ok(TimeoutStream { stream }))
     }
 }
 
 #[pin_project]
-pub struct TimeoutBody<B> {
+#[derive(Debug)]
+pub struct TimeoutStream<S> {
```

Comment:
What prevents us from being able to use `tokio_io_timeout::TimeoutStream` directly?

Reply:
We need to implement hyper's `Connection` trait, which the inner one doesn't do.